app/models/notification/purging.rb

```diff
@@ -0,0 +1,68 @@
+class Notification
+  module Purging
+    extend ActiveSupport::Concern
+    include PurgingMixin
+
+    module ClassMethods
+      def purge_date
+        nil
+      end
+
+      def purge_window_size
+        ::Settings.notifications.history.purge_window_size
+      end
+
+      # Collects the notifications that have expired based on the expires_in
+      # value for each notification type.  The "older_than" value is not used
+      # for this function, so it is named as `_`.
+      #
+      # First, this collects all of the notification_types and groups them by
+      # expires_in value (we only seed 3 different types currently).
+      #
+      # It then builds a single query that matches against each
+      # notification_type and if it has expired that specific expires_in for
+      # that type.
+      def purge_scope(_)
+        query = nil
+
+        # Make a query to collect the expires_in and id columns for all of the
+        # NotificationTypes, and then put them into a hash with the key being
+        # the expires_in column, and the values being all of the ids matching
+        # the expires_in value.
+        expire_ids = NotificationType.pluck(:expires_in, :id)
+        expire_ids = expire_ids.each_with_object({}) do |(key, id), result|
+          result[key] ||= []
+          result[key] << id
+        end
+
+        # Create a query that matches on the notification on expires_in for
+        # specific ids that have that exact same expires_in time. So something
+        # like:
+        #
+        #   SELECT *
+        #   FROM "notifications"
+        #   WHERE ((("notifications"."notification_type_id" IN (1,2,3)
+        #          AND "notifications"."created_at" < [[1.days.ago.utc]])
+        #      OR ("notifications"."notification_type_id" IN (4,5,6)
+        #          AND "notifications"."created_at" < [[7.days.ago.utc]]))
+        #      OR ("notifications"."notification_type_id" IN (7,8,9)
+        #          AND "notifications"."created_at" < [[14.days.ago.utc]]))
+        #
+        expire_ids.each do |expires_in, type_ids|
+          new_condition = arel_table.grouping(
+            arel_table[:notification_type_id].in(type_ids).and(
+              arel_table[:created_at].lt(expires_in.seconds.ago.utc)
+            )
+          )
+          query = query.nil? ? new_condition : query.or(new_condition)
+        end
```

Comment:
Maybe I'm misunderstanding, but why all of this complexity, when this is a relatively simple time based query?

Offhand (I didn't fully test this but I will; just wanted to post before this is merged)...

```ruby
Notification.all.joins(:notification_type).select("notifications.created_at + notification_types.expires_in * INTERVAL '1 second'")
```

should get you all notifications that are past expiry.

Now, granted, that may not use indexes, but I'm curious if that even matters.  However, that being said, I would argue we should add a "expires_on" column to the notification itself which, on creation, would do Time.now + notification_type.expires on.  Then the purge scope is a super simple `.where(:expires_on < Time.now.utc)`

Reply:
Tried this locally, and it worked (would obviously need more testing, but I'm confident enough to share)

```ruby
Notification.all.joins(:notification_type).where("notifications.created_at + notification_types.expires_in * INTERVAL '1 second' < now()").delete_all
```

Reply:
@Fryguy On PTO, but wanted to respond before it is off your mind:

> Offhand... [+ your query and stuff]

Well first, I think you are trying to do the "where" part in the the `.select`, which won't work... or filter on anything relevant (maybe my brain is also turned off because I have been not working today).

* * *

Second, I did try to do this in a single query, and only briefly mentioned it in the PR description since usually you don't read lengthy descriptions... but since you asked :trollface: 👍 

I don't think it is possible in SQL, or at least easily possible, but I might be limited in my knowledge with this.  Trying to filter on the `:created_at` while simultaneously calculate for each record what the expires date is based on the `:created_at` in the `.where`.  This is wrong, but this is what I was trying to do:

```SQL
SELECT *
FROM notifications
JOIN notification_types ON notifications.notification_type_id = notification_types.id
WHERE notifications.created_at < (now() - notification_types.expires_in * INTERVAL '1 second');
```

Seems like it would be a really bad query to run because this part:

```SQL
(now() - notification_types.expires_in * INTERVAL '1 second')
```

Would have to be run for each record, I think.  Again, I tried, but it seemed like just grabbing the interval types up front, sorting the uniq types of `expires_in`, and building this query was much less expensive then the above (and less complicated to put into ruby/rails).  I can try again on Monday though.

Reply:
Clearly you posted your second comment while I was still typing mine, but I will try the above monday (looks more sane because you aren't filtering in a `.select` :wink: )

Reply:
Since you are on PTO I forbid you to read this until you are back 😆 

> Seems like it would be a really bad query to run because this part ... would have to be run for each record

That is basically what database servers do anyway though when you don't have an index.  They create a hash-lookup of just the criteria from each record, then do some HashJoin on it.

Here's the EXPLAIN from that delete, FWIW.  I don't have a database with a lot of notifications though.  I would love to run all of these different queries against it with EXPLAIN ANALYZE and see what happens.  Perhaps we can get one if you don't have one already?

```text
=> EXPLAIN for: DELETE FROM "notifications" WHERE "notifications"."id" IN (SELECT "notifications"."id" FROM "notifications" INNER JOIN "notification_types" ON "notification_types"."id" = "notifications"."notification_type_id" WHERE (notifications.created_at + notification_types.expires_in * INTERVAL '1 second' < now()))
                                                            QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------------
Delete on notifications  (cost=21.16..32.10 rows=38 width=18)
  ->  Nested Loop  (cost=21.16..32.10 rows=38 width=18)
        ->  HashAggregate  (cost=21.02..21.40 rows=38 width=20)
              Group Key: notifications_1.id
              ->  Hash Join  (cost=2.15..20.92 rows=38 width=20)
                    Hash Cond: (notifications_1.notification_type_id = notification_types.id)
                    Join Filter: ((notifications_1.created_at + ((notification_types.expires_in)::double precision * '00:00:01'::interval)) < now())
                    ->  Seq Scan on notifications notifications_1  (cost=0.00..14.50 rows=450 width=30)
                    ->  Hash  (cost=1.51..1.51 rows=51 width=18)
                          ->  Seq Scan on notification_types  (cost=0.00..1.51 rows=51 width=18)
        ->  Index Scan using notifications_pkey on notifications  (cost=0.15..0.27 rows=1 width=14)
              Index Cond: (id = notifications_1.id)
```

Reply:
> Since you are on PTO I forbid you to read this until you are back

Confession, I might have skimmed this in my inbox, but refrained from actually acting on it.

* * *

So it's cool that what you were thinking was my initial first impression of what to do, I just didn't interpret the postgresql docs as well...

Tried this with my tests that I wrote as l well, and it seems to work.   Thanks for figuring out how to do this the right way!

> Note that seqscan in there, which may cause problems, but I'd like to see this with a real-world table.

Yeah, I want to put a script together to populate a DB with a large amount of notifications and do some benchmarks on both versions.  I really do like the simplicity of yours, but unsure if it is slower or not.  Will report back.

Reply:
Since the above will properly take a while, I am going to push up a WIP commit that just shows the difference for those passing through.